### PR TITLE
feat: Add temperature setting to `ilab model chat`

### PR DESCRIPTION
**Pull request by alinaryan**
Introduced a default temperature setting in the config, with a default of 1.0. Also includes a `--temperature` and `-t` flag to allow users to set a custom temperature for each chat session. This enables users to adjust the temperature value, leading to their preferred level of deterministic and accurate responses.

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Functional tests have been added, if necessary.
- [ ] E2E Workflow tests have been added, if necessary.

### CODE REVIEW

File: src/instructlab/configuration.py [hunk 1]
```diff
@@ -146,6 +146,10 @@ class _chat(BaseModel):
         default=None,
         description="The maximum number of tokens that can be generated in the chat completion. Be aware that larger values use more memory.",
     )
+    temperature: float = Field(
+        default=0.7,
```
Comment (RobotSail) [inline src/instructlab/configuration.py:150]: Is 0.7 what we always use as a default somewhere? It seems a bit high to have as the default. Have we tested what the output is like with this setting? We should make sure that this will give better results than other values.

**Review by RobotSail — COMMENTED** [review]
Thank you for adding this, it'll be very useful. Could you also please provide some examples for the chat output when we use 0.7? It seems pretty high to have as the default value. I'd expect a good default to be somewhere closer to the 0.1-0.3 range.

File: src/instructlab/model/chat.py [hunk 2]
```diff
@@ -173,6 +173,13 @@ def is_openai_server_and_serving_model(
     required=False,
     help="Log file path to write server logs to.",
 )
+@click.option(
+    "-temp",
```
Comment (alimaredia) [inline src/instructlab/model/chat.py:177]: generally options with one `-` just have one or two letters after them. I think a `-t` here would work.

File: src/instructlab/model/chat.py [hunk 3]
```diff
@@ -348,6 +358,7 @@ def __init__(
         self.log_file = log_file
         self.greedy_mode = greedy_mode
```
Comment (alimaredia) [inline src/instructlab/model/chat.py:359]: Can we get rid of greedy_mode entirely, or make a note to get rid of it next release. Being able to set the temperature in a flag and a config variable should be enough.

File: src/instructlab/model/chat.py [hunk 4]
```diff
@@ -635,11 +646,11 @@ def start_prompt(
             self.multiline_mode = 0
             self.multiline = not self.multiline
 
-        # Optional parameters
-        create_params = {}
-        if self.greedy_mode:
+        # Temperature parameters
```
Comment (alimaredia) [inline src/instructlab/model/chat.py:649]: It looks like the default temperature if `self.greedy_mode` was `False` here before? https://platform.openai.com/docs/api-reference/chat/create#chat-create-temperature

@RobotSail what would you think about making the default for temperature `None`? I think we'd want some number to be the default so that the user always knows what the temperature is?

File: src/instructlab/model/chat.py [hunk 5]
```diff
@@ -635,11 +646,11 @@ def start_prompt(
...
+        create_params = {
```
Comment (alimaredia) [inline src/instructlab/model/chat.py:650]: Nit: Initialize an empty dict for `create_params` first, and then add temperature for code readability purposes.

File: src/instructlab/model/chat.py (hunk 4 above)
Comment (RobotSail) [reply src/instructlab/model/chat.py:649]: @alimaredia For temperature we should never use `None`. It should be an explicit float between 0 & 1.

File: src/instructlab/configuration.py (hunk 1 above)
Comment (alimaredia) [reply src/instructlab/configuration.py:150]: @alinaryan @RobotSail am I reading the old code right. It seems if `greed_mode` was not set then `create_params['temperature']` is never set, which means the temperature is 1 which is the default here: https://platform.openai.com/docs/api-reference/chat/create#chat-create-temperature.

I know we had discussed what the default temperature value should be. It looks like it's `1.0` in vLLM as well (https://github.com/vllm-project/vllm/blob/main/vllm/sampling_params.py#L176) and `0.8` in llama-cpp (https://github.com/ggerganov/llama.cpp/blob/master/examples/main/README.md?plain=1#L170).

Maybe we just leave it at 1.0 then?

Comment (alinaryan) [reply src/instructlab/configuration.py:150]: Does 1.0 seem like it is too high? Let me experiment with chat and I'll post results on this PR or in slack

File: src/instructlab/model/chat.py [hunk 6]
```diff
@@ -174,6 +174,12 @@ def is_openai_server_and_serving_model(
     required=False,
     help="Log file path to write server logs to.",
 )
+@click.option(
+    "-t," "--temperature",
+    is_flag=True,
```
Comment (alimaredia) [inline src/instructlab/model/chat.py:179]: Do we need `is_flag` here?

File: src/instructlab/model/chat.py [hunk 7]
```diff
@@ -174,6 +174,12 @@ def is_openai_server_and_serving_model(
...
+    cls=clickext.ConfigOption,
+    required=True,  # default from config
```
Comment (alimaredia) [inline src/instructlab/model/chat.py:181]: I don't think we want this to be required

File: src/instructlab/model/chat.py [hunk 8]
```diff
@@ -183,7 +189,6 @@ def chat(
     context,
     session,
     quick_question,
-    greedy_mode,
```
Comment (alimaredia) [inline src/instructlab/model/chat.py:186]: The option for greedy_mode click option will need to be removed too

File: tests/test_config.py [hunk 9]
```diff
@@ -45,7 +45,6 @@ def _assert_defaults(self, cfg: config.Config):
         assert cfg.chat.context == "default"
         assert cfg.chat.session is None
         assert cfg.chat.logs_dir == f"{data_dir}/chatlogs"
-        assert not cfg.chat.greedy_mode
```
Comment (alimaredia) [inline tests/test_config.py:48]: Do we need to add a check here for temperature?

File: src/instructlab/configuration.py (hunk 1 above)
Comment (alinaryan) [reply src/instructlab/configuration.py:146]: @alimaredia and I spoke offline and agreed that the default should be set to 1.0, since that is the OpenAI default anyways. That default serves as a reinforcement to our default so it makes sense for them to be the same value.

File: src/instructlab/model/chat.py (hunk 8 above)
Comment (alinaryan) [reply src/instructlab/model/chat.py:186]: Removed this on lines 141-144

File: CHANGELOG.md [hunk 10]
```diff
@@ -10,6 +10,10 @@
 * `ilab` now supports system profiles. These profiles apply entire configuration files tailored to specific hardware configurations. We support a set of auto-detected profiles for CPU enabled Linux machines, M-Series Apple Silicon Chips, and Nvidia GPUs, and Intel Gaudi 3. When you run `ilab config init`, one of these profiles should be selected for you. If there is not a direct match, a menu will be displayed allowing you to choose one.
 * Add support for inferencing with IBM granite architecture models.
 
+* `ilab chat` now includes a temperature setting feature, allowing users to adjust the response generation behavior. A default temperature of 1.0
```
Comment (nathan-weinberg) [inline CHANGELOG.md:13]:
```suggestion
* `ilab model chat` now includes a temperature setting feature, allowing users to adjust the response generation behavior. A default temperature of 1.0
```

File: src/instructlab/train/linux_train.py [hunk 11]
```diff
@@ -297,7 +297,7 @@ def model_generate(user, **kwargs):
             input_ids=input_ids,
             max_new_tokens=256,
             pad_token_id=tokenizer.eos_token_id,
-            temperature=0.7,
+            temperature=1.0,
```
Comment (nathan-weinberg) [inline src/instructlab/train/linux_train.py:300]: Why this change?

File: tests/test_config.py (hunk 9 above)
Comment (nathan-weinberg) [reply tests/test_config.py:48]: This is done as part of the latest patchset

**Review by nathan-weinberg — COMMENTED** [review]
few minor comments but otherwise LGTM

can you add something in the CHANGELOG under `Breaking Changes` about greedy mode being deprecated?

File: src/instructlab/train/linux_train.py (hunk 11 above)
Comment (alinaryan) [reply src/instructlab/train/linux_train.py]: Spoke with @alimaredia, this change is not needed here

File: CHANGELOG.md (hunk 10 above)
Comment (alinaryan) [reply CHANGELOG.md]: Added, thanks

**Review by alimaredia — APPROVED** [review]

**Review by cdoern — APPROVED** [review]
I didn't know `randomness` was a word but it is :laughing:

**Review by nathan-weinberg — APPROVED** [review]
